src/pnpq/devices/odl_thorlabs_kbd101.py

```diff
@@ -24,138 +49,202 @@ def __init__(
             )
-        # relative
+        self.curr_pos_steps += steps
 
-        msg = b"\x48\x04\x06\x00\xd0\x01\x00\x00"
-        msg = msg + (int(steps)).to_bytes(4, byteorder="little", signed=True)
-        self.conn.write(msg)
+    def __ensure_final_in_range(self, current_position: int, steps: int):
 
-        forward_complete = self.waitForReply(b"\x64\x04", self.move_timeout)
-        # if not forward_complete:
-        #    raise Warning("Can not received STEP_FW Complete!")
-        # else:
-        #    print("Step Forward complete:" + str(forward_complete))
+        max_threshold = self.maxmove * self.resolution
+        min_threshold = self.maxmove * self.resolution
 
-    def step_backward(self, steps):
-        if not self.conn.is_open:
-            raise Exception(
-                "Move forward failed: can not connect to Thorlabs ODL device"
+        next_position = current_position + steps
+        if next_position > max_threshold or next_position < min_threshold:
+            raise OdlMoveOutofRangeError(
+                f"The relative change position request for device{self} is out of range min({self.minmove}) - max({self.maxmove})"
             )
 
-        if steps > self.maxmove * self.resolution:
-            raise Exception(
-                "required steps are more that the device resolution: "
-                + str(self.resolution)
-            )
+    def step_backward(self, steps):
+        self.__ensure_port_open()
+        self.__ensure_steps_in_range(steps)
 
         # negate steps
         steps = -steps
+        self.__ensure_final_in_range(self.curr_pos_steps, steps)
+
         # relative move
-        msg = b"\x48\x04\x06\x00\xd0\x01\x00\x00"
-        msg = msg + (int(steps)).to_bytes(4, byteorder="little", signed=True)
+        msg = ODL_RELATIVE_MOVE_COMMAND + (int(steps)).to_bytes(
+            4, byteorder="little", signed=True
+        )
         self.conn.write(msg)
 
-        forward_complete = self.waitForReply(b"\x64\x04", self.move_timeout)
-        # if not forward_complete:
-        #    raise Warning("Can not received STEP_FW Complete!")
-        # else:
-        #    print("Step Forward complete:" + str(forward_complete))
+        backward_result = self.__wait_for_reply(b"\x64\x04", self.move_timeout)
+        if not backward_result:
+            self.logger.error(f"step backward command is not completed")
+            raise OdlMoveNotCompleted(
+                f"ODL{self}: No response is received for step_backward command"
+            )
+        self.curr_pos_steps += steps
 
-    def get_status(self):
-        if not self.conn.is_open:
-            raise Exception("ODL get_status failed: can not connect to ODL device")
+    def auto_update_start(self) -> bytes | None:
+        self.logger.info("cal auto update start cmd")
+        self.__ensure_port_open()
+        msg = START_UPDATE_COMMAND
+        self.conn.write(msg)
```

Comment:
you don't need to assign the cmd to the `msg` variable. it should be `self.conn.write(START_UPDATE_COMMAND)` for readability.
if you use this kind of variable, you should use more informative name than the original.
in this case, the variable name `msg` loses the important information.